File: built-in plugins dsp/mb-util-testtone.cpp
/**
 * MB Test Tone Generator
 * Category : effect
 * Type     : eq
 * Version  : 1.0.0
 * Author   : Max Booster
 * Desc     : Generate test tones, sweeps, and noise for calibration
 *
 * Part of Max Booster Built-In Plugins DSP
 */

#ifndef MB_UTIL_TESTTONE_H
#define MB_UTIL_TESTTONE_H

#include <algorithm>
#include <cmath>
#include <cstring>
#include "PluginBase.h"

class MbUtilTesttone : public PluginBase {
public:
    static constexpr const char* PLUGIN_ID      = "mb-util-testtone";
    static constexpr const char* PLUGIN_NAME    = "MB Test Tone Generator";
    static constexpr const char* PLUGIN_TYPE    = "eq";
    static constexpr const char* PLUGIN_CATEGORY = "effect";
    static constexpr const char* VERSION         = "1.0.0";

    struct Parameters {
    float frequency = 1000f;  // range [20, 20000]
    float level = -20f;  // range [-60, 0]
    float waveform = 0f;  // range [0, 4]
    };

    MbUtilTesttone() = default;
    ~MbUtilTesttone() override = default;

    void setSampleRate(double sampleRate) override {
        sampleRate_ = sampleRate;
        reset();
    }

    void reset() override {
        std::memset(buffer_, 0, sizeof(buffer_));
    }

    void process(float* left, float* right, int numSamples, Parameters params) {
        params.frequency = std::clamp(params.frequency, 20f, 20000f);
        params.level = std::clamp(params.level, -60f, 0f);
        params.waveform = std::clamp(params.waveform, 0f, 4f);
        for (int i = 0; i < numSamples; ++i) {
            left[i]  = processSample(left[i],  params);
            right[i] = processSample(right[i], params);
        }
    }

private:
    double sampleRate_ = 44100.0;
    float  buffer_[65536] = {};

    inline float processSample(float input, const Parameters& params) {
        // DSP implementation for MB Test Tone Generator
        return input;
    }
};

#endif // MB_UTIL_TESTTONE_H

File: built-in plugins dsp/mb-util-testtone.ts
import type { PluginDefinition } from '../server/services/pluginHostService';

const MbUtilTesttonePlugin: PluginDefinition = { id: 'mb-util-testtone', slug: 'mb-util-testtone', name: 'MB Test Tone Generator', category: 'effect', type: 'eq' as any, version: '1.0.0', description: 'Generate test tones, sweeps, and noise for calibration', author: 'Max Booster', grade: 'A', parameters: [{ id: 'frequency', name: 'Frequency', type: 'float', defaultValue: 1000, minValue: 20, maxValue: 20000, automatable: true }, { id: 'level', name: 'Level', type: 'float', defaultValue: -20, minValue: -60, maxValue: 0, automatable: true }, { id: 'waveform', name: 'Waveform', type: 'float', defaultValue: 0, minValue: 0, maxValue: 4, automatable: false }], defaultPreset: { frequency: 1000, level: -20, waveform: 0 } };

export default MbUtilTesttonePlugin;
